ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/KafkaConfigStore.java

```diff
@@ -0,0 +1,203 @@
+    }
+  }
+
+  private Optional<KsqlProperties> readProperties(
+      final String topicName,
+      final Supplier<KafkaConsumer<String, KsqlProperties>> consumerSupplier) {
+    final TopicPartition topicPartition = new TopicPartition(topicName, 0);
+    final List<TopicPartition> topicPartitionAsList = Collections.singletonList(topicPartition);
+
+    try (KafkaConsumer<String, KsqlProperties> consumer = consumerSupplier.get()) {
+      consumer.assign(topicPartitionAsList);
+      consumer.seekToBeginning(topicPartitionAsList);
+
+      final Map<TopicPartition, Long> offsets = consumer.endOffsets(topicPartitionAsList);
+      final long endOffset = offsets.get(topicPartition);
+      while (consumer.position(topicPartition) < endOffset) {
+        log.debug(
+            "Reading from config topic. Position(%d) End(%d)",
+            consumer.position(topicPartition),
+            endOffset);
+        final ConsumerRecords<String, KsqlProperties> records
+            = consumer.poll(Duration.of(5, ChronoUnit.SECONDS));
+        final Optional<ConsumerRecord<String, KsqlProperties>> record =
+            records.records(topicPartition)
+                .stream()
+                .filter(r -> r.key().equals(CONFIG_MSG_KEY))
+                .findFirst();
+        if (record.isPresent()) {
+          log.info("Found existing config in config topic");
+          return Optional.of(record.get().value());
+        }
+      }
+      log.info("No config found on config topic");
+      return Optional.empty();
+    }
+  }
+
+  private void writeProperties(
+      final String topicName,
+      final KsqlProperties properties,
+      final Supplier<KafkaProducer<String, KsqlProperties>> producerSupplier) {
+    try (KafkaProducer<String, KsqlProperties> producer = producerSupplier.get()) {
+      producer.send(new ProducerRecord<>(topicName, CONFIG_MSG_KEY, properties));
+      producer.flush();
+    }
+  }
+
+  private Map<String, String> readMaybeWriteProperties(
+      final KsqlConfig ksqlConfig,
+      final KafkaTopicClient topicClient,
+      final Supplier<KafkaConsumer<String, KsqlProperties>> consumerSupplier,
+      final Supplier<KafkaProducer<String, KsqlProperties>> producerSupplier) {
+    final String topicName = KsqlInternalTopicUtils.getTopicName(ksqlConfig, CONFIG_TOPIC_SUFFIX);
+    log.info("Ensure existence of topic %s", topicName);
+    KsqlInternalTopicUtils.ensureTopic(
+        topicName, ksqlConfig, topicClient, TopicConfig.CLEANUP_POLICY_COMPACT);
+
+    final Optional<KsqlProperties> properties = readProperties(topicName, consumerSupplier);
+    if (properties.isPresent()) {
+      return properties.get().getKsqlProperties();
+    }
+
+    log.info("Writing current config to config topic");
+    writeProperties(topicName, new KsqlProperties(
+        ksqlConfig.getAllConfigPropsWithSecretsObfuscated()), producerSupplier);
+
+    return readProperties(topicName, consumerSupplier).get().getKsqlProperties();
```

Comment:
readProperties() may return an Optional.empty() which would cause .get() to throw an Exception, right?

Reply:
Nar, the fact that we've written properties on the line above means there's no way this can return `empty`.

Reply:
Ah, that's true.

Another question on this code, why not  returning the newly KsqlProperties created instead of reading it from the topic again? The writeProperties should fail if it cannot write to the topic, if it succeeds, then is there a reason to read it again?

Reply:
If we assume that all the servers are writing the config (which is a reasonable assumption), then yeah. That said, if we read it again we do get some defense from a user misconfiguring.

Reply:
Cool, thanks.